lib/python/Jetson/GPIO/gpio_event.py

```diff
@@ -48,195 +66,287 @@
+        
+            event_data = cdev.gpioevent_data.from_buffer_copy(data)
+
+            # event result
+            if event_data.id == cdev.GPIOEVENT_REQUEST_RISING_EDGE:
+                print("GPIOEVENT_REQUEST_RISING_EDGE")
+            elif event_data.id == cdev.GPIOEVENT_REQUEST_FALLING_EDGE:
+                print("GPIOEVENT_REQUEST_FALLING_EDGE")
+            else:
+                print("unknown event")
+                continue
+                   
+            _mutex.acquire()
             # check key to make sure gpio object has not been deleted
             # from main thread
-            key = _get_gpio_obj_key(fd)
-            if key is None:
-                continue
-
-            gpio_obj = _gpio_event_list[key]
+            chip_name, pin_num = _get_gpio_obj_keys(fd)
+            if channel != pin_num:
+                warnings.warn("Channel does not match with assigned file descriptor ", RuntimeWarning)
+                # _thread_running = False
+                _thread_running_dict[thread.get_ident()] = False
+                break
+            
+            # Get the gpio object to do following updates
+            gpio_obj = _get_gpio_object(chip_name, pin_num)
+            if gpio_obj is None:
+                raise RuntimeError("GPIO object does not exists")
+
+            # debounce the input event for the specified bouncetime
+            time = datetime.now()
+            time = time.second * 1E6 + time.microsecond
+            if (gpio_obj.bouncetime is None or
+                    (time - gpio_obj.lastcall >
+                        gpio_obj.bouncetime * 1000) or
+                    (gpio_obj.lastcall == 0) or gpio_obj.lastcall > time):
+                gpio_obj.lastcall = time
+                gpio_obj.event_occurred = True
 
-            # ignore first epoll trigger
-            if gpio_obj.initial_thread:
-                gpio_obj.initial_thread = False
-                _gpio_event_list[key] = gpio_obj
-            else:
-                # debounce the input event for the specified bouncetime
-                time = datetime.now()
-                time = time.second * 1E6 + time.microsecond
-                if (gpio_obj.bouncetime is None or
-                        (time - gpio_obj.lastcall >
-                         gpio_obj.bouncetime * 1000) or
-                        (gpio_obj.lastcall == 0) or gpio_obj.lastcall > time):
-                    gpio_obj.lastcall = time
-                    gpio_obj.event_occurred = True
-                    _gpio_event_list[key] = gpio_obj
-                    _mutex.release()
-                    for cb_func in gpio_obj.callbacks:
-                        cb_func()
+                #update to the original list
+                _gpio_event_list[chip_name][pin_num] = gpio_obj
+                _mutex.release()
 
+                # callback function
+                for cb_func in gpio_obj.callbacks:
+                    cb_func()
```

Comment:
Let's change the callback function so that it receive the time stamp and whether it is a falling or rising edge event